cloudsync/smartsync.py

```diff
@@ -0,0 +1,435 @@
+            ent[REMOTE].sync_path = None
+            ent[REMOTE].sync_hash = None
+        self.requestset.discard(ent)
+        self.excludeset.add(ent)
+
+    def _smart_unsync(self, ents, source_id) -> Optional[SyncEntry]:
+        if not ents:
+            raise ex.CloudFileNotFoundError(source_id)
+        for ent in ents:
+            if ent in self.requestset:
+                self._smart_unsync_ent(ent)
+                return ent
+        return None
+
+    def smart_unsync_ent(self, ent) -> Optional[SyncEntry]:
+        return self._smart_unsync([ent], ent)
+
+    def smart_unsync_oid(self, remote_oid) -> Optional[SyncEntry]:
+        ent = self.lookup_oid(REMOTE, remote_oid)
+        return self._smart_unsync([ent], remote_oid)
+
+    def smart_listdir_path(self, side, path):
+        """returns the ents for all files in remote folder by looking in the state db, doesn't hit the provider api."""
+        r_prov = self.providers[side]
+        for ent, _ignored_rel_path in self.get_kids(path, side):
+            if r_prov.paths_match(r_prov.dirname(ent[side].path), path):
+                yield ent
+
+    @property
+    def changes(self):
+        # Filter the changes to just those that are in the requestset, or those whose get_latest is stale viz. change_time
+        changes = [ent[LOCAL].path for ent in super().changes]
+        for ent in self.requestset:
+            if ent[LOCAL].path in changes:
+                yield ent
+
+    @property
+    def _changeset(self):
+        # this implies a remote walk on startup, otherwise autosynced items won't autosync until an event shows up
+        # do we want to do the autosync calculation on every do()? Or perhaps flag each entry that the
+        #     autosync calculation has been done, and no need to do it again... if so, don't persist that flag
+        changes = set()
+        for ent in self._changeset_storage:
+            if ent in self.excludeset and not ent[LOCAL].changed:
+                continue
+
+            included = False
+            if ent in self.requestset:
+                included = True
+            elif ent[REMOTE].otype == DIRECTORY:
+                included = True  # simplifies syncing files, avoids needing to sync the parent(s) later
+            elif (ent[REMOTE].changed or ent[LOCAL].changed) and not ent.is_latest():
+                included = True  # needs a get_latest() at least
+            elif not ent[LOCAL].oid:  # this means the entry is not currently synced locally
+                for callback in self._callbacks:
+                    if ent[REMOTE].path and callback(ent[REMOTE].path):
+                        self._smart_sync_ent(ent)
+                        included = True
+                        break
+
+            if included:
+                changes.add(ent)
+
+        return changes
+
+    @_changeset.setter
+    def _changeset(self, value):
+        self._changeset_storage = value
+
+
+@dataclass
+class FSInfo(OInfo):
```

Comment:
is there a good reason why this is "FS" info, when it's not, necessarily, a filesystem?   it could be an oracle database.   please just add a "custom" field to oinfo.